src/main/java/com/amplitude/api/AmplitudeClient.java

```diff
@@ -392,39 +394,25 @@ public AmplitudeClient useAdvertisingIdForDeviceId() {
      * Enable location listening in the SDK. This will add the user's current lat/lon coordinates
      * to every event logged.
      *
+     * This function should be called before SDK initialization, e.g. {@link #initialize(Context, String)}.
+     *
      * @return the AmplitudeClient
      */
     public AmplitudeClient enableLocationListening() {
-        runOnLogThread(new Runnable() {
-            @Override
-            public void run() {
-                if (deviceInfo == null) {
-		    throw new IllegalStateException(
-		            "Must initialize before acting on location listening.");
-                }
-                deviceInfo.setLocationListening(true);
-            }
-        });
+        this.locationListening = true;
```

Comment:
don't you need to update locationListening flag in deviceInfo

Reply:
Since it only fetches once and then caches the device info, even you have flag set, it won't read it again.

Also this function needs to call before initialization before DeviceInfo fetches the location info.

Reply:
In logEvent, it always asks DeviceInfo for latest location: https://github.com/amplitude/Amplitude-Android/blob/master/src/main/java/com/amplitude/api/AmplitudeClient.java#L1054.

You might want to check the locationListening flag in AmplitudeClient first before fetching latest location.

Also if the customer disables location listening, but wants to re-enable it, we should probably support that by also updating flag inside Device Info

Reply:
Good catch! Thank you

Reply:
Updated PR.

Reply:
@haoliu-amp I forgot to mention that this behavior is only in Android SDK. The iOS SDK fetches the location only once, and caches it, and sends it on all events, whereas Android SDK fetches the most recent location every event

Reply:
You mean cached location info inside DeviceInfo?

Reply:
👍